src/Shared/CertificateGeneration/UnixCertificateManager.cs

```diff
@@ -47,26 +130,707 @@ protected override X509Certificate2 SaveCertificateCore(X509Certificate2 certifi
+    /// so we should handle the same extensions as c_rehash (other than .crl).
+    /// </remarks>
+    [GeneratedRegex("\\.(pem|crt|cer)$")]
+    private static partial Regex OpenSslCertificateExtensionRegex();
+
+    /// <remarks>
+    /// This is a simplified version of c_rehash from OpenSSL.  Using the real one would require
+    /// installing the OpenSSL perl tools and perl itself, which might be annoying in a container.
+    /// </remarks>
+    private static bool TryRehashOpenSslCertificates(string certificateDirectory)
+    {
+        try
+        {
+            // First, delete all the existing symlinks, so we don't have to worry about fragmentation or leaks.
+
+            var hashRegex = OpenSslHashFilenameRegex();
+            var extensionRegex = OpenSslCertificateExtensionRegex();
+
+            var certs = new List<FileInfo>();
+
+            var dirInfo = new DirectoryInfo(certificateDirectory);
+            foreach (var file in dirInfo.EnumerateFiles())
+            {
+                var isSymlink = (file.Attributes & FileAttributes.ReparsePoint) == FileAttributes.ReparsePoint;
+                if (isSymlink && hashRegex.IsMatch(file.Name))
+                {
+                    file.Delete();
+                }
+                else if (extensionRegex.IsMatch(file.Name))
+                {
+                    certs.Add(file);
+                }
+            }
+
+            // Then, enumerate all certificates - there will usually be zero or one.
+
+            // c_rehash doesn't create additional symlinks for certs with the same fingerprint,
+            // but we don't expect this to happen, so we favor slightly slower look-ups when it
+            // does, rather than slightly slower rehashing when it doesn't.
+
+            foreach (var cert in certs)
+            {
+                if (!TryGetOpenSslHash(cert.FullName, out var hash))
+                {
+                    return false;
+                }
+
+                var linkCreated = false;
+                for (var i = 0; i < MaxHashCollisions; i++)
+                {
+                    var linkPath = Path.Combine(certificateDirectory, $"{hash}.{i}");
+                    if (!File.Exists(linkPath))
+                    {
+                        // As in c_rehash, we link using a relative path.
+                        File.CreateSymbolicLink(linkPath, cert.Name);
+                        linkCreated = true;
+                        break;
+                    }
+                }
+
+                if (!linkCreated)
+                {
+                    Log.UnixOpenSslRehashTooManyHashes(cert.FullName, hash, MaxHashCollisions);
+                    return false;
+                }
+            }
+        }
+        catch (Exception ex)
+        {
+            Log.UnixOpenSslRehashException(ex.Message);
+            return false;
+        }
+
+        return true;
+    }
```

Comment:
I'm not sure our tool should change trust for other certificates that aren't our own. We should check that the cert we are going to trust is an https dev cert and skip it otherwise.

Reply:
Can you please elaborate?  I started with logic that just manipulated certs and links in 1:1 pairs, but then learned that there are also collisions and switched to the purge-and-rehash approach used by c_rehash.

Reply:
Yes, there can be collisions, but I believe it's fine, the behavior is to append `.0`, `.1` and so on, so iterate over until you find the first free one.

The reason I suggest this is because it's way less invasive to do this only for our certs than to apply this to all the certs in the folder, potentially breaking unrelated certs.

A mitigating factor for this is if we have high confidence that this only runs against our specific folder, in which case, all the certs in the folder should be ours, is that the case?

Reply:
If we wanted to go "as simple as possible", the hash that gets generated for the certificate is based on the distinguished subject name (CN=localhost) which means that all the dev-certs (present and past) are going to have the same symbolic link name (with the `.0`, `.1`, and so on). 

So, we could alternatively to doing all this, look in the directory for files until we find the next valid suffix and create the symbolic link for that file.

Reply:
> Yes, there can be collisions, but I believe it's fine, the behavior is to append .0, .1 and so on, so iterate over until you find the first free one.

I was concerned that the consuming code might not like gaps in the list.  Re-hashing the directory has the benefit of defragging the sequence.

> The reason I suggest this is because it's way less invasive to do this only for our certs than to apply this to all the certs in the folder, potentially breaking unrelated certs.

It should have the same effect as running `c_rehash`.  If the user's file layout can't survive that, they're already taking a risk.

> A mitigating factor for this is if we have high confidence that this only runs against our specific folder, in which case, all the certs in the folder should be ours, is that the case?

Yes, it's our own specific folder.  Of course, users can discover it and insert their own certs there.

> the hash that gets generated for the certificate is based on the distinguished subject name (CN=localhost)

I don't have any knowledge about what gets hashed, but I find this surprising since the subject name doesn't seem to contain enough data to make a SHA-1 hash meaningful.  Having said that, it does make sense that you'd want to look up certs by subject.

> So, we could alternatively to doing all this, look in the directory for files until we find the next valid suffix and create the symbolic link for that file.

Mimicking `c_rehash` feels more forward compatible and less surprising to consumers (which we don't control).

Reply:
> -subject_hash
> Prints the "hash" of the certificate subject name. This is used in OpenSSL to form an index to allow certificates in a directory to be looked up by subject name.

https://docs.openssl.org/3.3/man1/openssl-x509/#certificate-printing-options

Reply:
At that point, we could presumably treat it as a constant, given that our subject name is a constant.  Not shelling out seems like a nice win, but I think I'd rather handle that in a follow-up PR so we can separately discuss the trade-offs of the optimization (e.g. it will no longer pick up certs the user puts in our magic directory).

Reply:
> e.g. it will no longer pick up certs the user puts in our magic directory

Should that be a non-goal? If users put their own certs in that folder, we shouldn't do anything with them, they should not rely on our tool to set up trust for their own handcrafted certs? They can run `c_rehash` if they want to, we don't have to support and maintain that do we?

> At that point, we could presumably treat it as a constant, given that our subject name is a constant. Not shelling out seems like a nice win, but I think I'd rather handle that in a follow-up PR so we can separately discuss the trade-offs of the optimization

Sound fair, my goal here is simply that we have "less code" to maintain moving forward.

> Yes, it's our own specific folder. Of course, users can discover it and insert their own certs there.

We should discourage people from doing that, the same way we discourage them from using our OID.

Reply:
> Should that be a non-goal? If users put their own certs in that folder, we shouldn't do anything with them, they should not rely on our tool to set up trust for their own handcrafted certs? They can run c_rehash if they want to, we don't have to support and maintain that do we?

I don't feel strongly either way - I just thought we should explicitly make the decision.

Reply:
I leave it up to you, however, in my view, if we aren't sure, we shouldn't do it. It's easier to add it later than remove it if we need to.

Reply:
I plan to put up a new PR scaling back the behavior this week - I'd just like to merge this large PR in a state I feel confident I understand.  Plus, it will give us something to resuscitate if we decide we want the richer behavior.